src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,171 @@
+import RequestQueue from "./RequestQueue";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Resolver = (value?: any) => void;
+type Rejecter = (reason?: unknown) => void;
+
+type RequestSubscription = {
+  subscriberId: number;
+  resolve: Resolver;
+  reject: Rejecter;
+};
+
+/**
+ * An extension of `RequestQueue` that adds a concept of "subscribers," which may share references to a single request
+ * or cancel their subscription without disrupting the request for other subscribers.
+ */
+export default class SubscribableRequestQueue {
+  private queue: RequestQueue;
+
+  /** Number of subscribers, used for generating unique subscriber IDs. */
+  private subscriberIds: number;
+  /** Map keyed by subscriber ID. Subscribers are only useful for cancelling early, so we only store rejecters here. */
+  private subscribers: Map<number, Map<string, Rejecter>>;
+  /** Map from "inner" request (managed by `queue`) to "outer" promises generated per-subscriber. */
+  private requests: Map<string, RequestSubscription[]>;
+
+  constructor(maxActiveRequests?: number) {
+    this.queue = new RequestQueue(maxActiveRequests);
+    this.subscriberIds = 0;
+    this.subscribers = new Map();
+    this.requests = new Map();
+  }
+
+  /** Resolves all subscriptions to request `key` with `value */
+  private resolveAll<T>(key: string, value: T) {
+    const requests = this.requests.get(key);
+    if (requests) {
+      for (const { resolve, subscriberId } of requests) {
+        resolve(value);
+        this.subscribers.get(subscriberId)?.delete(key);
+      }
+      this.requests.delete(key);
+    }
+  }
+
+  /** Rejects all subscriptions to request `key` with `reason` */
+  private rejectAll(key: string, reason: unknown) {
+    const requests = this.requests.get(key);
+    if (requests) {
+      for (const { reject, subscriberId } of requests) {
+        reject(reason);
+        this.subscribers.get(subscriberId)?.delete(key);
+      }
+      this.requests.delete(key);
+    }
+  }
+
+  /** Adds a new request subscriber. Returns a unique ID to identify this subscriber. */
+  addSubscriber(): number {
+    const subscriberId = this.subscriberIds;
+    this.subscriberIds++;
+    this.subscribers.set(subscriberId, new Map());
+    return subscriberId;
+  }
+
+  /** Queues a new request, or adds a subscription if the request is already queued/running. */
+  addRequestToQueue<T>(
+    key: string,
+    subscriberId: number,
+    requestAction: () => Promise<T>,
+    delayMs?: number
+  ): Promise<T> {
+    // Create single underlying request if it does not yet exist
+    if (!this.queue.hasRequest(key)) {
+      this.queue
+        .addRequest(key, requestAction, delayMs)
+        .then((value) => this.resolveAll(key, value))
+        .catch((reason) => this.rejectAll(key, reason));
+    }
+    if (!this.requests.has(key)) {
+      this.requests.set(key, []);
+    }
```

Comment:
Would there ever be an instance where the keys of `queue` and `request` become desynced?

Reply:
(I don't think there would be, looking at the code? This just feels funny to me but doesn't do anything bad)

Reply:
Yeah, we should never see only one or the other of these `if`s evaluate to true, but I still separated them for extra robustness.